# urllib的使用
是python3自带的内置模块
包含的方法：
- urllib.request:打开和读取URL
- urllib.error:包含urllib.request抛出的异常
- urllib.parse:用于解析URL
- urllib.robotparser:用于解析robots.txt

## urllib的网址请求

In [3]:
import urllib.request
response = urllib.request.urlopen('https://cn.bing.com/?mkt=zh-CN')
print(response.read().decode('utf-8')) #调用read方法可以得到返回的网页内容并打印网页代码

<!doctype html><html lang="zh" dir="ltr"><head><meta name="theme-color" content="#4F4F4F" /><meta http-equiv="X-UA-Compatible" content="IE=edge" /><meta name="viewport" content="width=device-width, initial-scale=1.0" /><title>必应</title><link rel="preconnect" href="https://r.bing.com" /><link rel="preconnect" href="https://r.bing.com" crossorigin/><link rel="dns-prefetch" href="https://r.bing.com" /><link rel="dns-prefetch" href="https://r.bing.com" crossorigin/><link rel="stylesheet" href="/rp/M98Jfa5O8Kvq0pTB2mCGZrlTeCM.css" type="text/css"/><script type="text/javascript">//<![CDATA[
var logJSText=function(n,t){t===void 0&&(t=null);(new Image).src=_G.lsUrl+'&Type=Event.ClientInst&DATA=[{"T":"CI.ClientInst","FID":"CI","Name":"'+escape(n)+(t?'","Text":"'+escape(t):"")+'"}]'},getHref=function(){return location.href};try{var ignErr=["ResizeObserver loop completed with undelivered notifications.","ResizeObserver loop limit exceeded"],rszMsg="ResizeObserverLimit",maxErr=3,ignoreCurrentError

如果网速比较慢或者所请求的网站打开较慢，可以设置一个超时限制，加入timeout参数

In [4]:
import urllib.request
response = urllib.request.urlopen('https://cn.bing.com/?mkt=zh-CN',timeout=8)
print(response.read().decode('utf-8')) #使用read方法读取网页，decode设置编码方式

<!doctype html><html lang="zh" dir="ltr"><head><meta name="theme-color" content="#4F4F4F" /><meta http-equiv="X-UA-Compatible" content="IE=edge" /><meta name="viewport" content="width=device-width, initial-scale=1.0" /><title>必应</title><link rel="preconnect" href="https://r.bing.com" /><link rel="preconnect" href="https://r.bing.com" crossorigin/><link rel="dns-prefetch" href="https://r.bing.com" /><link rel="dns-prefetch" href="https://r.bing.com" crossorigin/><link rel="stylesheet" href="/rp/M98Jfa5O8Kvq0pTB2mCGZrlTeCM.css" type="text/css"/><script type="text/javascript">//<![CDATA[
var logJSText=function(n,t){t===void 0&&(t=null);(new Image).src=_G.lsUrl+'&Type=Event.ClientInst&DATA=[{"T":"CI.ClientInst","FID":"CI","Name":"'+escape(n)+(t?'","Text":"'+escape(t):"")+'"}]'},getHref=function(){return location.href};try{var ignErr=["ResizeObserver loop completed with undelivered notifications.","ResizeObserver loop limit exceeded"],rszMsg="ResizeObserverLimit",maxErr=3,ignoreCurrentError

## 网页的保存和异常处理

In [6]:
try:
    x = urllib.request.urlopen('https://cn.bing.com/?mkt=zh-CN')
    #print(x.read)
    save = open('1.html','w')
    save.write(str(x.read()))
    save.close()
except Exception as e:
    print(str(e))

## 构造请求对象Requests
使用Requests模仿浏览器，避免一些简单的反爬措施，"远古的方法"

In [8]:
import urllib.request
request = urllib.request.Request('https://cn.bing.com/?mkt=zh-CN') # 用Request构造Request对象类
response = urllib.request.urlopen(request)
print(response.read().decode('utf-8'))

<!doctype html><html lang="zh" dir="ltr"><head><meta name="theme-color" content="#4F4F4F" /><meta http-equiv="X-UA-Compatible" content="IE=edge" /><meta name="viewport" content="width=device-width, initial-scale=1.0" /><title>必应</title><link rel="preconnect" href="https://r.bing.com" /><link rel="preconnect" href="https://r.bing.com" crossorigin/><link rel="dns-prefetch" href="https://r.bing.com" /><link rel="dns-prefetch" href="https://r.bing.com" crossorigin/><link rel="stylesheet" href="/rp/M98Jfa5O8Kvq0pTB2mCGZrlTeCM.css" type="text/css"/><script type="text/javascript">//<![CDATA[
var logJSText=function(n,t){t===void 0&&(t=null);(new Image).src=_G.lsUrl+'&Type=Event.ClientInst&DATA=[{"T":"CI.ClientInst","FID":"CI","Name":"'+escape(n)+(t?'","Text":"'+escape(t):"")+'"}]'},getHref=function(){return location.href};try{var ignErr=["ResizeObserver loop completed with undelivered notifications.","ResizeObserver loop limit exceeded"],rszMsg="ResizeObserverLimit",maxErr=3,ignoreCurrentError

## 添加请求头
基本格式：
`headers = {'User-Agent':'请求头'}`
请求头，随便从一个网页上进行检查页面寻找，点击网络（network），查看标头
<img src="2.1_1.png" width="50%" height="50%"/>
* 请求网址：当前打开的网页
* 代码方法：200 OK，表示请求打开网址成功
* cookies：用于用户身份识别
* User-Agent：用户代理
<img src="2.1_2.png" width="50%" height="50%"/>

In [11]:
from urllib import request
url = 'https://www.zhihu.com/hot/'
headers = {'User-Agent':'Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Mobile Safari/537.36 Edg/121.0.0.0'}
req = request.Request(url=url,headers=headers)
response = request.urlopen(req)
print(response.read().decode('utf-8'))

<!DOCTYPE html>
<html lang="zh">
<head>
<meta charset="UTF-8" />
<meta name="viewport" content="width=device-width, initial-scale=1.0" />
<meta http-equiv="X-UA-Compatible" content="ie=edge,chrome=1" />
<meta name="renderer" content="webkit" />
<meta name="force-rendering" content="webkit" />
<title>404 - 知乎</title>
<link
rel="shortcut icon"
href="//static.zhihu.com/static/img/favicon.ico"
/>
<style>
body,
h1,
p,
pre {
margin: 0;
}
body {
font-family: -apple-system, BlinkMacSystemFont, "Helvetica Neue",
"PingFang SC", "Microsoft YaHei", "Source Han Sans SC",
"Noto Sans CJK SC", "WenQuanYi Micro Hei", sans-serif;
}
a {
text-decoration: none;
}
</style>
<style>
.Button {
background: none;
border: none;
cursor: pointer;
display: inline-block;
font-size: 14px;
padding: 0;
text-align: center;
}
.Button:focus {
outline: none;
transition: box-shadow 0.3s;
}
/* 暂时没有用 focus visiable */
.Button:focus {
box-shadow: 0 0 0 2px #ffffff, 0 0 0 4px rgba(0, 132, 255, 0.3);
}
html[data-theme="dark"] .Bu

在请求头中加入cookie，便于身份识别

In [13]:
import urllib
import urllib.request
url = 'https://m.jd.com/'
headers = {'User-Agent':'Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Mobile Safari/537.36 Edg/121.0.0.0',
          'Connection':'keep-alive',
          'Accept-Encoding':'',
          'Cookies':'unpl=JF8EAL5nAGQiaEhRURxSHRQZSl9VWVhYHxIKbmNXUlhYQlUHGQQeGxMZbVdfXg1XFgBzZwdFXV9AVAYQAhoiEEptVF9cCkwWCmpmDWRtWExQAU8FEkYVSjNUMFVxDW0DJ2d0EVw5e1cFKwMrEhZKW11eXQBJFwBvZgVUXlBKUQ0SBB0iEXtdXFptOHsSA25mBFxaWUxVNRoyGRMRTllVWVwLTCdIAWZIVFtZTV0FGwoZEhNLXFReXgBKEgtmYQNkXGhI;__jda=76161171.17078119312101129049877.1707811931.1707811931.1707811932.1; __jdc=76161171; __jdu=17078119312101129049877; 3AB9D23F7A4B3CSS=jdd03AACG7GM3N27GSRLU37EI45Z5CL4CYQOS22UWZCYRF2OYBIVKEQWFCBJBDI3LEKYYGTM3MENJLJXQLKDIGSYIRS6FFAAAAAMNUGDAKYAAAAAADMDXAC5MATFLUIX; 3AB9D23F7A4B3C9B=AACG7GM3N27GSRLU37EI45Z5CL4CYQOS22UWZCYRF2OYBIVKEQWFCBJBDI3LEKYYGTM3MENJLJXQLKDIGSYIRS6FFA; _gia_d=1; areaId=16; ipLoc-djd=16-1362-0-0; shshshfpa=941c3b14-072a-17cb-db38-8ec394a706da-1707811933; shshshfpx=941c3b14-072a-17cb-db38-8ec394a706da-1707811933; shshshsID=a2333f1175cf971d1e50a0cbacb28862_1_1707811933295; shshshfpb=BApXeuJyOouhA6pUE9W35f2nx57QvYbcABkXAHhxo9xJ1MuAf1oO2; __jdb=76161171.2.17078119312101129049877|1.1707811932; __jdv=76161171|www.baidu.com|t_1003608409_|tuiguang|24e6a77803064ae4805c74180337482c|1707811948228'}
request = urllib.request.Request(url=url,headers=headers)
# 发起请求
response = urllib.request.urlopen(request)
# 用read()方法读取文件里的全部内容，返回字符串
html = response.read().decode('utf-8')
# 返回HTTP的响应码，200表示成功
print(response.getcode())
# 返回数据实际URL，防止重新定向问题
print(response.geturl())
# 返回服务器响应的HTTP标头
print(response.info())
# 打印响应，返回内容
print(html)

200
https://m.jd.com/
Date: Tue, 13 Feb 2024 09:01:08 GMT
Content-Type: text/plain;charset=UTF-8
Content-Length: 37186
Connection: close
Vary: Accept-Encoding
Expires: Tue, 13 Feb 2024 09:01:08 GMT
Cache-Control: max-age=0
Server: jfe
Strict-Transport-Security: max-age=7776000


<!DOCTYPE html><html lang="zh-CN"><head>
  <meta charset="utf-8">
  <script type="text/javascript">window.alert=console.log;</script><script>
window._PFM_TIMING = [[1351, new Date()]];</script>
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0">
  <title>多快好省，购物逛京东！</title>
  <meta http-equiv="Cache-Control" content="max-age=180">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  <meta name="format-detection" content="telephone=no">
  <link rel="dns-prefetch" href="//m.360buyimg.com">
  <meta name="format-detection" content="telephone=no">
  <meta name="viewport" content="wid

## SSL认证
有SSL认证的网页一般是安全的，如果遇到没有SSL认证的网页可以添加SSL忽略警告而继续访问

In [14]:
import urllib
import urllib.request
# 对SSL进行设置，忽略警告，继续进行访问’
import ssl
ssl._create_default_https_context_ = ssl._create_unverified_context
url = 'https://www.12306.cn/mormhweb/'
response = urllib.request.urlopen(url=url)
print(response.read().decode('utf-8'))

<!DOCTYPE html>
<html lang="zh-CN">

<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
  <title>中国铁路12306网站</title>
  <script>
    window.startTime = new Date().getTime(); //"window.onload外开始时间:",window.startTime
  </script>
  <link rel="shortcut icon" href="images/favicon.ico" type="image/x-icon" />
  <!-- <link href="./css/index.css" rel="stylesheet">
  <link href="./css/global.css" rel="stylesheet">
  <link href="./css/public.css" rel="stylesheet"> -->
  <link href="./fonts/iconfont.css" rel="stylesheet">
  <!-- 日期城市控件 -->
  <!-- <link href="./css/common/calendarNew.css" rel="stylesheet">
  <link href="./css/common/table.css" rel="stylesheet">
  <link href="./css/common/station.css" rel="stylesheet">
  <link href="./css/common/station_new.css" rel="stylesheet">
  <link rel="stylesheet" href="./css/main.css"> -->
  <link rel="stylesheet" href="./css/index_y_v50002.css">
  <link href="./css/festival2024.css" rel="stylesheet">
</head>
<sty

这里也可以强制https认证（使用Requests请求）

In [16]:
import requests
# urllib3官方强制认证https安全证书，用于解决警告
import urllib3
urllib3.disable_warnings()
url = 'https://www.12306.cn/mormhweb/'
# requests操作简单，False表示不要进行认证
response = requests.get(url=url,verify=False)
# 设置返回编码方式
response.encoding='utf-8'
print(response.text)

<!DOCTYPE html>
<html lang="zh-CN">

<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
  <title>中国铁路12306网站</title>
  <script>
    window.startTime = new Date().getTime(); //"window.onload外开始时间:",window.startTime
  </script>
  <link rel="shortcut icon" href="images/favicon.ico" type="image/x-icon" />
  <!-- <link href="./css/index.css" rel="stylesheet">
  <link href="./css/global.css" rel="stylesheet">
  <link href="./css/public.css" rel="stylesheet"> -->
  <link href="./fonts/iconfont.css" rel="stylesheet">
  <!-- 日期城市控件 -->
  <!-- <link href="./css/common/calendarNew.css" rel="stylesheet">
  <link href="./css/common/table.css" rel="stylesheet">
  <link href="./css/common/station.css" rel="stylesheet">
  <link href="./css/common/station_new.css" rel="stylesheet">
  <link rel="stylesheet" href="./css/main.css"> -->
  <link rel="stylesheet" href="./css/index_y_v50002.css">
  <link href="./css/festival2024.css" rel="stylesheet">
</head>
<sty

# 万能视频下载
you-get模块(需要在Pycharm上使用)，参考官方文档[You-Get](https://you-get.org/)

In [ ]:
# 以下代码有待考究，在pycharm上也报错
import sys
from you_get import common as yg # 导入you-get库
# 设置下载目录
directory = r'存储地址'
# 设置要下载的视频地址
url = '视频链接'
sys.argv=['you-get','-o',directory,'--format=flv',url]
yg.main()

# Requests中get的使用
urllib实际使用不多，Requests使用比较多，因为其速度快

## Requests基础：代码获取

In [19]:
import requests
r = requests.get('https://cn.bing.com/?mkt=zh-CN')
print(r.text)

<!doctype html><html lang="zh" dir="ltr"><head><meta name="theme-color" content="#4F4F4F" /><meta http-equiv="X-UA-Compatible" content="IE=edge" /><meta name="viewport" content="width=device-width, initial-scale=1.0" /><title>必应</title><link rel="preconnect" href="https://r.bing.com" /><link rel="preconnect" href="https://r.bing.com" crossorigin/><link rel="dns-prefetch" href="https://r.bing.com" /><link rel="dns-prefetch" href="https://r.bing.com" crossorigin/><link rel="stylesheet" href="/rp/MkFOO6PSiaDl9oMqZycFBnxWoEc.br.css" type="text/css"/><script type="text/javascript">//<![CDATA[
var logJSText=function(n,t){t===void 0&&(t=null);(new Image).src=_G.lsUrl+'&Type=Event.ClientInst&DATA=[{"T":"CI.ClientInst","FID":"CI","Name":"'+escape(n)+(t?'","Text":"'+escape(t):"")+'"}]'},getHref=function(){return location.href};try{var ignErr=["ResizeObserver loop completed with undelivered notifications.","ResizeObserver loop limit exceeded"],rszMsg="ResizeObserverLimit",maxErr=3,ignoreCurrentEr

status_code方法可以获取到请求结果的状态

In [27]:
import requests
r = requests.get('https://blog.csdn.net/weixin_46211269')
print(r.status_code) # 打印状态码
print(r.text) # 打印文本

200



## Requests基础：构建请求

In [28]:
import requests
r = requests.get('https://httpbin.org/get')
print(r.text)

{
  "args": {}, 
  "headers": {
    "Accept": "*/*", 
    "Accept-Encoding": "gzip, deflate, br", 
    "Host": "httpbin.org", 
    "User-Agent": "python-requests/2.31.0", 
    "X-Amzn-Trace-Id": "Root=1-65cb501a-403c066d086680cd67c69b3f"
  }, 
  "origin": "185.248.85.36", 
  "url": "https://httpbin.org/get"
}



其中返回的参数：
+ origin：个人IP
+ url：请求的网址

## 获取Cookie
cookie属性（前两个是必须）：  
1.cookie的名字：name  
2.cookie的值：value  
3.cookie过期时间：Expires/Max-Age  
4.cookie作用路径：Path  
5.cookie所在名域：Domain  
6.使用cookie进行安全连接：Securre  

In [29]:
import requests
headers = {'User-Agent':'Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Mobile Safari/537.36 Edg/121.0.0.0'}
url = 'https://www.csdn.net/?spm=1011.2124.3001.5359'
r = requests.get(url=url,headers=headers)
cookiejar = r.cookies
cookiedict = requests.utils.dict_from_cookiejar(cookiejar)
print(cookiejar)
print(cookiedict)

<RequestsCookieJar[<Cookie https_waf_cookie=9a282351-4d80-4a5e7254b949e11856c708d46d62243e5944 for www.csdn.net/>, <Cookie uuid_tt_dd=10_37296770860-1707823805894-156647 for .csdn.net/>, <Cookie dc_session_id=10_1707823805894.483999 for .csdn.net/>]>
{'https_waf_cookie': '9a282351-4d80-4a5e7254b949e11856c708d46d62243e5944', 'uuid_tt_dd': '10_37296770860-1707823805894-156647', 'dc_session_id': '10_1707823805894.483999'}


## 添加请求头

In [32]:
# 未添加请求头，按理说是出现403Forbidden，表明被反爬
import requests
r = requests.get('https://www.zhihu.com/explore')
print(r.text)

<!doctype html>
<html lang="zh" data-hairline="true" class="itcauecng" data-theme="light"><head><meta charSet="utf-8"/><title data-rh="true">发现 - 知乎</title><meta name="viewport" content="width=device-width,initial-scale=1,maximum-scale=1"/><meta name="renderer" content="webkit"/><meta name="force-rendering" content="webkit"/><meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1"/><meta name="google-site-verification" content="FTeR0c8arOPKh8c5DYh_9uu98_zJbaWw53J-Sch9MTg"/><meta name="description" property="og:description" content="知乎，中文互联网高质量的问答社区和创作者聚集的原创内容平台，于 2011 年 1 月正式上线，以「让人们更好的分享知识、经验和见解，找到自己的解答」为品牌使命。知乎凭借认真、专业、友善的社区氛围、独特的产品机制以及结构化和易获得的优质内容，聚集了中文互联网科技、商业、影视、时尚、文化等领域最具创造力的人群，已成为综合性、全品类、在诸多领域具有关键影响力的知识分享社区和创作者聚集的原创内容平台，建立起了以社区驱动的内容变现商业模式。"/><link data-rh="true" rel="apple-touch-icon" href="https://static.zhihu.com/heifetz/assets/apple-touch-icon-152.81060cab.png"/><link data-rh="true" rel="apple-touch-icon" href="https://static.zhihu.com/heifetz/assets/apple-touch-icon-152.

In [33]:
# 添加请求头
import requests
headers = {'User-Agent':'Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Mobile Safari/537.36 Edg/121.0.0.0'}
r = requests.get("https://www.zhihu.com/explore",headers=headers)
print(r.text)

<!doctype html>
<html lang="zh" data-hairline="true" data-android="true" class="itcauecng" data-theme="light"><head><meta charSet="utf-8"/><title data-rh="true">发现 - 知乎</title><meta name="viewport" content="width=device-width,initial-scale=1,maximum-scale=1,user-scalable=0,viewport-fit=cover"/><meta name="renderer" content="webkit"/><meta name="force-rendering" content="webkit"/><meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1"/><meta name="google-site-verification" content="FTeR0c8arOPKh8c5DYh_9uu98_zJbaWw53J-Sch9MTg"/><meta name="description" property="og:description" content="知乎，中文互联网高质量的问答社区和创作者聚集的原创内容平台，于 2011 年 1 月正式上线，以「让人们更好的分享知识、经验和见解，找到自己的解答」为品牌使命。知乎凭借认真、专业、友善的社区氛围、独特的产品机制以及结构化和易获得的优质内容，聚集了中文互联网科技、商业、影视、时尚、文化等领域最具创造力的人群，已成为综合性、全品类、在诸多领域具有关键影响力的知识分享社区和创作者聚集的原创内容平台，建立起了以社区驱动的内容变现商业模式。"/><link data-rh="true" rel="apple-touch-icon" href="https://static.zhihu.com/heifetz/assets/apple-touch-icon-152.81060cab.png"/><link data-rh="true" rel="apple-touch-icon" href="https:

## 二进制数据获取

In [36]:
import requests
headers = {'User-Agent':'Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Mobile Safari/537.36 Edg/121.0.0.0'}
r = requests.get("https://gifhub.com/favicon.ico",headers=headers)
print(r.text)
print(r.content)


b''


In [38]:
import requests
headers = {'User-Agent':'Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Mobile Safari/537.36 Edg/121.0.0.0'}
r = requests.get("https://gifhub.com/favicon.ico",headers=headers)
with open("favicon.ico",'wb') as f:
    f.write(r.content)

爬取书本作者个人博客头像并保存

In [42]:
import requests
headers={
    'User-Agent':'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.11 (KHTML, like Gecko) Chrome/23.0.1271.64 Safari/537.11'}
url="https://profile-avatar.csdnimg.cn/171da61cd5e74ebfa9ab08e29de51894_weixin_46211269.jpg!1" 
r = requests.get(url=url,headers=headers)
with open("photo.jpg",'wb') as f:
    f.write(r.content)

- 注意：二进制图片保存时，open函数使用wb参数

# Requests中post的使用
除了get请求，还有Post，post的作用是将要处理的数据提交给服务器  
优点如下：  
1.比GET更安全  
2.可以上传的数据更大，可以使用POST发送文本数据和二进制数据  
缺点如下：  
1.Post请求不会被缓存  
2.POST请求不会被保留在浏览器历史记录中  
3.POST发送的数据在URL中不可见，因此无法为特定查询的页面添加书签。  

## 提交数据表单

In [43]:
import requests
data = {'name':'zyzyzyzy','age':'20'}
r = requests.post("https://httpbin.org/post",data=data)
print(r.text)

{
  "args": {}, 
  "data": "", 
  "files": {}, 
  "form": {
    "age": "20", 
    "name": "zyzyzyzy"
  }, 
  "headers": {
    "Accept": "*/*", 
    "Accept-Encoding": "gzip, deflate, br", 
    "Content-Length": "20", 
    "Content-Type": "application/x-www-form-urlencoded", 
    "Host": "httpbin.org", 
    "User-Agent": "python-requests/2.31.0", 
    "X-Amzn-Trace-Id": "Root=1-65cb6144-52aa12bf5bef2b507a8f536e"
  }, 
  "json": null, 
  "origin": "185.248.85.36", 
  "url": "https://httpbin.org/post"
}



## 添加请求图

In [44]:
import requests
import json
host = "https://httpbin.org/"
endpoint = "post"
url = ''.join([host,endpoint])
headers = {'User-Agent':'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.11 (KHTML, like Gecko) Chrome/23.0.1271.64 Safari/537.11'}
r = requests.post(url,headers=headers)
print(r.text)

{
  "args": {}, 
  "data": "", 
  "files": {}, 
  "form": {}, 
  "headers": {
    "Accept": "*/*", 
    "Accept-Encoding": "gzip, deflate, br", 
    "Content-Length": "0", 
    "Host": "httpbin.org", 
    "User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.11 (KHTML, like Gecko) Chrome/23.0.1271.64 Safari/537.11", 
    "X-Amzn-Trace-Id": "Root=1-65cb6b29-6bcff94709d0e6880c55c76e"
  }, 
  "json": null, 
  "origin": "185.248.85.36", 
  "url": "https://httpbin.org/post"
}



## 提交json

In [55]:
import requests
import json
host = "https://httpbin.org/"
endpoint = "post"
url = '/'.join([host, endpoint])
data = {
    "sites":[
        {"name":"zyzyzy","url":"https://www.icourse163.org/"},
        {"name":"pypypy","url":"https://space.bilibili.com/"},
        {"name":"fzufzu","url":"https://jwch.fzu.edu.cn/"}]}
r = requests.post(url,json=data)
if r.status_code == 200:
    try:
        response = r.json()
        print(response)
    except json.JSONDecodeError:
        print("解析JSON时出错。")
else:
    print(f"请求失败，状态码：{r.status_code}")

{'args': {}, 'data': '{"sites": [{"name": "zyzyzy", "url": "https://www.icourse163.org/"}, {"name": "pypypy", "url": "https://space.bilibili.com/"}, {"name": "fzufzu", "url": "https://jwch.fzu.edu.cn/"}]}', 'files': {}, 'form': {}, 'headers': {'Accept': '*/*', 'Accept-Encoding': 'gzip, deflate, br', 'Content-Length': '182', 'Content-Type': 'application/json', 'Host': 'httpbin.org', 'User-Agent': 'python-requests/2.31.0', 'X-Amzn-Trace-Id': 'Root=1-65cb70d2-5a1005ff18b4439d6e2d978c'}, 'json': {'sites': [{'name': 'zyzyzy', 'url': 'https://www.icourse163.org/'}, {'name': 'pypypy', 'url': 'https://space.bilibili.com/'}, {'name': 'fzufzu', 'url': 'https://jwch.fzu.edu.cn/'}]}, 'origin': '185.248.85.36', 'url': 'https://httpbin.org/post'}


## 普通文件上传

In [60]:
import requests
import json
host = "https://httpbin.org/"
endpoint = "post"
url = "".join([host,endpoint])
# 普通上传
files = {"file":open("text.txt",'rb')}
r = requests.post(url,files=files)
print(r.text)

{
  "args": {}, 
  "data": "", 
  "files": {
    "file": "\u798f\u5dde\u5927\u5b66\u672c\u79d1\u751f\u5956\u5b66\u91d1\u7ba1\u7406\u529e\u6cd5\r\n\u798f\u5927\u5b66\u30142018\u301556\u53f7\r\n\u7b2c\u4e00\u7ae0  \u603b  \u5219\r\n\u7b2c\u4e00\u6761  \u4e3a\u5168\u9762\u8d2f\u5f7b\u515a\u7684\u6559\u80b2\u65b9\u9488\uff0c\u9f13\u52b1\u5b66\u751f\u52e4\u594b\u5b66\u4e60\u3001\u79ef\u6781\u8fdb\u53d6\u3001\u52c7\u4e8e\u521b\u65b0\uff0c\u4fc3\u8fdb\u5b66\u751f\u5fb7\u3001\u667a\u3001\u4f53\u5168\u9762\u53d1\u5c55\uff0c\u57f9\u517b\u5408\u683c\u7684\u793e\u4f1a\u4e3b\u4e49\u4e8b\u4e1a\u7684\u5efa\u8bbe\u8005\u548c\u63a5\u73ed\u4eba\uff0c\u7ed3\u5408\u6211\u6821\u5b9e\u9645\u60c5\u51b5\uff0c\u7279\u5236\u5b9a\u672c\u529e\u6cd5\u3002\r\n\u7b2c\u4e8c\u6761  \u672c\u529e\u6cd5\u4e2d\u7684\u201c\u5b66\u751f\u201d\u662f\u6307\u5728\u6211\u6821\u63a5\u53d7\u666e\u901a\u9ad8\u7b49\u5b66\u5386\u6559\u80b2\u7684\u5168\u65e5\u5236\u672c\u79d1\u751f\u3002\r\n\u7b2c\u4e8c\u7ae0  \u5956\u5b66\u91d1\u7684

# Request进阶

## URLError和HTTPError
URLError一般有几种情况：
- 没有网络连接
- 服务器连接失败
- 找不到指定服务器
- try except捕获异常

In [61]:
import urllib
import urllib.request
url = "https://www.12306.cn/index/"
try:
    response = urllib.request.urlopen(url=url,timeout=5)
    print(response.read().decode('utf-8'))
except Exception as err:
    print(err)

<!DOCTYPE html>
<html lang="zh-CN">

<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
  <title>中国铁路12306网站</title>
  <script>
    window.startTime = new Date().getTime(); //"window.onload外开始时间:",window.startTime
  </script>
  <link rel="shortcut icon" href="images/favicon.ico" type="image/x-icon" />
  <!-- <link href="./css/index.css" rel="stylesheet">
  <link href="./css/global.css" rel="stylesheet">
  <link href="./css/public.css" rel="stylesheet"> -->
  <link href="./fonts/iconfont.css" rel="stylesheet">
  <!-- 日期城市控件 -->
  <!-- <link href="./css/common/calendarNew.css" rel="stylesheet">
  <link href="./css/common/table.css" rel="stylesheet">
  <link href="./css/common/station.css" rel="stylesheet">
  <link href="./css/common/station_new.css" rel="stylesheet">
  <link rel="stylesheet" href="./css/main.css"> -->
  <link rel="stylesheet" href="./css/index_y_v50002.css">
  <link href="./css/festival2024.css" rel="stylesheet">
</head>
<sty

HTTPError是在Web开发和网络请求中遇到的一种常见错误，通常由requests库或其他HTTP客户端库抛出，指示HTTP请求失败。HTTPError错误的发生通常与HTTP响应状态码有关，表示服务器返回了一个错误状态。以下是一些可能导致HTTPError的常见情况，分别对应不同的HTTP状态码：
1. 4xx 客户端错误
- 400 Bad Request：请求无效或无法被服务器理解。通常是发送的请求数据有误。
- 401 Unauthorized：请求未经授权。通常需要通过适当的验证头部进行身份验证。
- 403 Forbidden：服务器拒绝执行此请求。与401不同，403错误表示服务器理解请求但拒绝授权。
- 404 Not Found：服务器找不到请求的资源。最常见的情况，表示URL可能写错了。
- 405 Method Not Allowed：请求的方法（GET、POST、DELETE等）对指定的资源不适用。
2. 5xx 服务器错误
- 500 Internal Server Error：服务器遇到了一个未知的错误，无法完成客户端的请求。
- 502 Bad Gateway：作为网关或代理工作的服务器从上游服务器收到了无效响应。
- 503 Service Unavailable：服务器目前无法使用（由于超载或停机维护）。通常，这只是暂时状态。
- 504 Gateway Timeout：作为网关或代理的服务器未及时从上游服务器接收请求。
3. 其他情况
- 网络连接问题：客户端与服务器之间的网络连接出现问题，如DNS解析失败，网络中断，或连接超时，也可能导致请求失败，尽管这不一定总是抛出HTTPError。
- 重定向过多：客户端在尝试追踪重定向时超过了最大重定向次数，虽然这可能会引发TooManyRedirects异常，而不是HTTPError。

## Fiddler的下载与简单使用
[Fiddler下载网址](https://www.telerik.com/download/fiddler)

[使用教程1：【B站精选fiddler抓包教程，学完别去做坏事！】 ](https://www.bilibili.com/video/BV1vM4y1V78M/?share_source=copy_web&vd_source=6d8b84248830d39f76ffdea73bf29974)  
[使用教程2：【2023自学fiddler抓包，请一定要看完【如何1天学会fiddler抓包】的全网最详细视频教程！！】]( https://www.bilibili.com/video/BV1o54y1u7UU/?share_source=copy_web&vd_source=6d8b84248830d39f76ffdea73bf29974)

## 登录网站

In [1]:
import urllib
import urllib.request

# 微博个人主页
url = 'https://weibo.com/u/7786677600'
headers={
    'User-Agent':'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36 Edg/121.0.0.0',
    'Cookie':'SINAGLOBAL=1932959923162.2695.1707835697374; ariaDefaultTheme=default; ariaFixed=true; ariaReadtype=1; ariaMouseten=null; ariaStatus=false; XSRF-TOKEN=HKN1WYD2nuBFm3od9bBk6upv; PC_TOKEN=5e503f438a; login_sid_t=67eac92d0ca79c21bd0a35a8b2f10d95; cross_origin_proto=SSL; WBStorage=267ec170|undefined; _s_tentry=weibo.com; wb_view_log=1707*10671.5; Apache=5110847105503.498.1707876254959; ULV=1707876254963:2:2:2:5110847105503.498.1707876254959:1707835697376; ALF=1710468270; SUB=_2A25IyG_-DeRhGeFJ41QX9ynKyzyIHXVrpO02rDV8PUJbkNB-LXDwkW1NfuaeXy0r3qOtil2hHrXSgCt2aO94fEc0; SUBP=0033WrSXqPxfM725Ws9jqgMF55529P9D9WhO7DmKdU4Pi6AGKjf-Z6ZX5JpX5o275NHD95QNS0ncSoMNSo57Ws4DqcjMi--NiK.Xi-2Ri--ciKnRi-zNS0MRSoqNS0q7e5tt; WBPSESS=Dt2hbAUaXfkVprjyrAZT_GaDtsOTngHT-mkSZdtnYK0X-CSulKZArMaqfbfbprJddm98gywlsmdcIYmGaKiS3sZP1VvZ1FJidXadp0Qof0fZrU2ph7IRi0d1eiSPsn_s-8PgzftaSK-Fv0lo5iJBp985WEXuGs15mel9gAa6IU1ZJLVzHwwuoiEVdXn3WnxzBGW-Il_AkANpRCE4tY4Ppg=='
}
# 模仿浏览器获取数据
request = urllib.request.Request(url=url,headers=headers)
# 发起一个请求
response = urllib.request.urlopen(request)
print(response.read().decode('utf-8'))

<!DOCTYPE html><html lang=zh-cn><head><meta charset=utf-8><link rel=dns-prefetch href=//h5.sinaimg.cn><meta name=viewport content="width=device-width,initial-scale=1,user-scalable=no,viewport-fit=cover"><meta http-equiv=X-UA-Compatible content="IE=edge"><meta http-equiv=Content-Security-Policy content=upgrade-insecure-requests><link rel=icon href=https://weibo.com/favicon.ico><title>微博</title><script>if (window.location.protocol !== 'https:') {
        window.location.href = window.location.href.replace('http:', 'https:');
      }</script><link href=//h5.sinaimg.cn/m/weibo-pro/css/chunk-00e28620.032d9cc6.css rel=prefetch><link href=//h5.sinaimg.cn/m/weibo-pro/css/chunk-019705cd.d9eea5c4.css rel=prefetch><link href=//h5.sinaimg.cn/m/weibo-pro/css/chunk-04dd4069.47a5a2b0.css rel=prefetch><link href=//h5.sinaimg.cn/m/weibo-pro/css/chunk-074455b5.cc383c83.css rel=prefetch><link href=//h5.sinaimg.cn/m/weibo-pro/css/chunk-0b2f6aca.2c8e2b71.css rel=prefetch><link href=//h5.sinaimg.cn/m/weibo-

Cookie在爬虫方面的典型应用是判定注册用户是否已经登录网站，用户可能会得到提示，是否在下一次进入网站时保留用户信息，以便简化登录手续

## 代理设置
对于大规模且频繁的请求网站可能会弹出验证码，或者跳转到登录验证界面，也可能会直接封禁客户端IP，导致永久无法访问。  
可以通过设置代理来解决----->用到proxies参数

In [15]:
import requests
proxies = {
    'http':'https://117.57.92.247:8089',
    'http':'http://183.164.243.37:8089'
}
response = requests.get('https://cn.bing.com/?mkt=zh-CN',proxies=proxies)
print(response.text)

<!doctype html><html lang="zh" dir="ltr"><head><meta name="theme-color" content="#4F4F4F" /><meta http-equiv="X-UA-Compatible" content="IE=edge" /><meta name="viewport" content="width=device-width, initial-scale=1.0" /><title>必应</title><link rel="preconnect" href="https://r.bing.com" /><link rel="preconnect" href="https://r.bing.com" crossorigin/><link rel="dns-prefetch" href="https://r.bing.com" /><link rel="dns-prefetch" href="https://r.bing.com" crossorigin/><link rel="stylesheet" href="/rp/MkFOO6PSiaDl9oMqZycFBnxWoEc.br.css" type="text/css"/><script type="text/javascript">//<![CDATA[
var logJSText=function(n,t){t===void 0&&(t=null);(new Image).src=_G.lsUrl+'&Type=Event.ClientInst&DATA=[{"T":"CI.ClientInst","FID":"CI","Name":"'+escape(n)+(t?'","Text":"'+escape(t):"")+'"}]'},getHref=function(){return location.href};try{var ignErr=["ResizeObserver loop completed with undelivered notifications.","ResizeObserver loop limit exceeded"],rszMsg="ResizeObserverLimit",maxErr=3,ignoreCurrentEr

上述代码的proxies里面应该要有https，但是没有找到合适的代理IP  
查找[代理IP的网站](https://ip.ihuan.me/)  
上述网站的IP不一定可用

# 实战演练

## 获取某搜索的代码

In [16]:
import requests
url = 'https://cn.bing.com/?mkt=zh-CN'
r = requests.get(url)
print(r.text)

<!doctype html><html lang="zh" dir="ltr"><head><meta name="theme-color" content="#4F4F4F" /><meta http-equiv="X-UA-Compatible" content="IE=edge" /><meta name="viewport" content="width=device-width, initial-scale=1.0" /><title>必应</title><link rel="preconnect" href="https://r.bing.com" /><link rel="preconnect" href="https://r.bing.com" crossorigin/><link rel="dns-prefetch" href="https://r.bing.com" /><link rel="dns-prefetch" href="https://r.bing.com" crossorigin/><link rel="stylesheet" href="/rp/MkFOO6PSiaDl9oMqZycFBnxWoEc.br.css" type="text/css"/><script type="text/javascript">//<![CDATA[
var logJSText=function(n,t){t===void 0&&(t=null);(new Image).src=_G.lsUrl+'&Type=Event.ClientInst&DATA=[{"T":"CI.ClientInst","FID":"CI","Name":"'+escape(n)+(t?'","Text":"'+escape(t):"")+'"}]'},getHref=function(){return location.href};try{var ignErr=["ResizeObserver loop completed with undelivered notifications.","ResizeObserver loop limit exceeded"],rszMsg="ResizeObserverLimit",maxErr=3,ignoreCurrentEr

## 下载图片到本地

In [20]:
import requests
src = "https://tvax1.sinaimg.cn/crop.0.0.1080.1080.180/0067bQddly1hd4ui233erj30u00u03zh.jpg?KID=imgbed,tva&Expires=1707890676&ssig=gVixdbPg17"
r = requests.get(src)
with open('yonex.jpg','wb') as f:
    f.write(r.content)
print('下载完成')

下载完成


## 下载视频到本地

In [32]:
import requests
src = '视频链接'
r = requests.get(src,verify=False)
with open('movie.mp4', 'wb') as f:
    f.write(r.content)
print('下载完成')

下载完成


上述代码实测并不可行！？？

## 爬取翻译网站
[百度翻译](https://fanyi.baidu.com/)  
1. 输入翻译内容，检查网页
2. 发现翻译都在sug文件中（点预览）
3. 查看标头
<img src="fanyi.png" width='50%' height='50%'/>

点击“标头”可以发现请求网址：https://fanyi.baidu.com/sug ,请求方式：POST  
点击“响应”可以发现数据类型：字典，k：需要翻译的内容，v：翻译后的内容

In [39]:
import json
import requests
url = 'https://fanyi.baidu.com/sug'
word = input("请输入要翻译的单词或短语：")
data = {'kw':word}
headers = {'User-Agent':'Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Mobile Safari/537.36 Edg/121.0.0.0'}
response = requests.post(url=url,data=data,headers=headers)
dic = response.json()
# print(dic)
filename = word+'.json'
with open(filename,'w',encoding='utf-8') as fp:
    json.dump(dic,fp=fp,ensure_ascii=False)
if dic['data']:
    j = dic['data'][0]['v']
    print(j)
else:
    print('数据列表为空或不存在该索引')

请输入要翻译的单词或短语：晚上好
good evening
